### PR TITLE
 fix: pull in @loganfsmyth's SourceMap fixes

### PR DESCRIPTION
**Pull request by bcoe**
This pulls in @loganfsmyth's modifications to the remapping algorithm for SourceMaps.

I have a couple questions that I've left inline, please note that the tests that need to be modified reference the following:

```js
  // Original source
  // export class SimpleClass {
  //     hy() {
  //         console.log("Hy");
  //     }
  // }
  // Transpiled Source
  // "use strict";
  // var SimpleClass = (function () {
  //     function SimpleClass() {
  //     }
  //     SimpleClass.prototype.hy = function () {
  //         console.log("Hy");
  //     };
  //     return SimpleClass;
  // }());
  // exports.SimpleClass = SimpleClass;
```

CC: @loganfsmyth, @SimenB

### CODE REVIEW

File: packages/istanbul-lib-source-maps/test/map-store.test.js [hunk 1]
```diff
@@ -36,7 +36,7 @@ describe('map store', function () {
         },
         "end": {
           "line": 4,
-          "column": 5
+          "column": Infinity
```
Comment (bcoe) [inline packages/istanbul-lib-source-maps/test/map-store.test.js]: @loganfsmyth is it weird that line 4 (in a 5 line file) extends to the column position `Infinity`? Your comment indicates that this edge-case should only occur when we've hit the end of the file.

File: packages/istanbul-lib-source-maps/test/map-store.test.js [hunk 2]
```diff
@@ -79,8 +79,8 @@ describe('map store', function () {
             "column": 0
           },
           "end": {
-            "line": null,
-            "column": -1
+            "line": 1,
+            "column": 13
```
Comment (bcoe) [inline packages/istanbul-lib-source-maps/test/map-store.test.js]: It seemed a bit weird that the `decl` mapping for `(anonymous_1)` suddenly started mapping to something; what is `(anonymous_1)`? I think this might be referring to `export class` as if it represents a function?

File: packages/istanbul-lib-source-maps/test/map-store.test.js [hunk 3]
```diff
@@ -102,8 +102,8 @@ describe('map store', function () {
             "column": 4
           },
           "end": {
-            "line": null,
-            "column": -1
+            "line": 2,
```
Comment (bcoe) [inline packages/istanbul-lib-source-maps/test/map-store.test.js]: similarly it seemed weird that `(anonymous_3)` had a different `decl` which includes the entire function body, as `decl` is supposed to only include the class declaration I believe, so: `function foo ()`.

This could be the fault of Istanbul though, and how code is transpiled?

```js
  // var SimpleClass = (function () {
  //     function SimpleClass() {
  //     }
  //     SimpleClass.prototype.hy = function () {
  //         console.log("Hy");
  //     };
  //     return SimpleClass;
  // }());
```

**Comment by bcoe** [comment]
@coreyfarrell would love your eye balls on this.

File: packages/istanbul-lib-source-maps/lib/transformer.js [hunk 4]
```diff
@@ -13,24 +13,88 @@ function isInvalidPosition (pos) {
...
+    // Convert that original position back to a generated one, with a bump
+    // to the right, and a rightward bias. Since 'generatedPositionFor' searches
+    // for mappings in the original-order sorted list, this will find the
+    // mapping that corresponds to the one immediately after the
+    // beforeEndMapping mapping.
+    var afterEndMapping = sourceMap.generatedPositionFor({
+        source: beforeEndMapping.source,
+        line: beforeEndMapping.line,
+        column: beforeEndMapping.column + 1,
+        bias: 2
```
Comment (loganfsmyth) [inline packages/istanbul-lib-source-maps/lib/transformer.js]: I had this because the existing code does it already, but FYI the bias values are also exposed as constants:
```
bias: SourceMapConsumer.LEAST_UPPER_BOUND
```
if you want.

File: packages/istanbul-lib-source-maps/test/map-store.test.js (hunk 1 above)
Comment (loganfsmyth) [reply packages/istanbul-lib-source-maps/test/map-store.test.js]: The difficulty is that sourcemap mappings don't actually contain the end position data, so all you know is that it extends to the end of a given line, with no sense of what column that might be. You'd have to fetch the source content and split it by line to know the actual column offset.

I agree though that it may well require some downstream validation. I guess you could always just do `Number.MAX_SAFE_INTEGER` if you wanted to avoid dangers with possible `Infinity` values or something, but hard to say if that's actually any less likely to have issues.

File: packages/istanbul-lib-source-maps/test/map-store.test.js (hunk 2 above)
Comment (loganfsmyth) [reply packages/istanbul-lib-source-maps/test/map-store.test.js]: I'm having a hard time figuring out how this data blob works, so not sure I can add anything.

File: packages/istanbul-lib-source-maps/test/map-store.test.js (hunk 1 above)
Comment (coreyfarrell) [reply packages/istanbul-lib-source-maps/test/map-store.test.js]: I haven't done a full review yet (I have other commitments today), but I just wanted to comment that I did check `Infinity` against the second argument to `String.slice`, `String.substr` and `String.substring`.  All worked as expected.

I'm could be convinced if someone could show a way that `Infinity` could break something but I'd prefer we did not use `Number.MAX_SAFE_INTEGER`.  `Infinity` is clearly not a specific number where `Number.MAX_SAFE_INTEGER` is.

Comment (bcoe) [reply packages/istanbul-lib-source-maps/test/map-store.test.js]: the fact that `Infinity` behaves nicely with the String helpers is good to know. We're talking about making this a breaking change any ways, and I'm feeling more and more confident about swapping out what were already undefined ranges with `Infinity1.

File: packages/istanbul-lib-source-maps/test/map-store.test.js (hunk 2 above)
Comment (bcoe) [reply packages/istanbul-lib-source-maps/test/map-store.test.js]: @loganfsmyth is it possible we're treating the `export` or `class` snippet as a `function` token type in the AST, and this is what `(anonymous_1)` refers to? The new offsets seem better than `null`, and `-1` ... which seems like a bug with a test around it, no?

@coreyfarrell does this set off any alarm bells for you?

File: packages/istanbul-lib-source-maps/test/map-store.test.js (hunk 1 above)
Comment (coreyfarrell) [reply packages/istanbul-lib-source-maps/test/map-store.test.js]: Granted `undefined` also works with String helpers but does not work with comparisons.
```js
assert(5 < Infinity); // pass
assert(5 < undefined); // fail
```

So I think this is 1) an improvement and 2) appropriate to flag as a breaking change just in case.

File: packages/istanbul-lib-source-maps/test/map-store.test.js (hunk 2 above)
Comment (coreyfarrell) [reply packages/istanbul-lib-source-maps/test/map-store.test.js]: The `(anonymous_1)` is referring to the IIFE.

So we've instrumented the transpiled source rather than instrumenting the original source.  I'm not fully sure how this is supposed to work, on my own projects I put instrumentation before other babel transformations.  I realize that could probably be a problem with TypeScript (the code added by instrumentation may not pass TypeScript checks).
